crates/nu-plugin/src/plugin/interface/stream_data_io.rs

```diff
@@ -0,0 +1,402 @@
+    ///
+    /// Other streams will be transparently handled or stored for concurrent readers.
+    fn read_list(&self) -> Result<Option<Value>, ShellError>;
+
+    /// Read some bytes for an `ExternalStream`'s `stdout` stream, returning `Ok(None)` at end
+    /// of stream.
+    ///
+    /// Other streams will be transparently handled or stored for concurrent readers.
+    fn read_external_stdout(&self) -> Result<Option<Vec<u8>>, ShellError>;
+
+    /// Read some bytes for an `ExternalStream`'s `stderr` stream, returning `Ok(None)` at end
+    /// of stream.
+    ///
+    /// Other streams will be transparently handled or stored for concurrent readers.
+    fn read_external_stderr(&self) -> Result<Option<Vec<u8>>, ShellError>;
+
+    /// Read a value for an `ExternalStream`'s `exit_code` stream, returning `Ok(None)` at end
+    /// of stream.
+    ///
+    /// Other streams will be transparently handled or stored for concurrent readers.
+    fn read_external_exit_code(&self) -> Result<Option<Value>, ShellError>;
+
+    /// Signal that no more values are desired from a `ListStream` and further messages should
+    /// be ignored.
+    fn drop_list(&self);
+
+    /// Signal that no more bytes are desired from an `ExternalStream`'s `stdout` and further
+    /// messages should be ignored.
+    fn drop_external_stdout(&self);
+
+    /// Signal that no more bytes are desired from an `ExternalStream`'s `stderr` and further
+    /// messages should be ignored.
+    fn drop_external_stderr(&self);
+
+    /// Signal that no more values are desired from an `ExternalStream`'s `exit_code` and further
+    /// messages should be ignored.
+    fn drop_external_exit_code(&self);
+
+    /// Write a value for a `ListStream`, or `None` to signal end of stream.
+    fn write_list(&self, value: Option<Value>) -> Result<(), ShellError>;
+
+    /// Write some bytes for an `ExternalStream`'s `stdout` stream, or `None` to signal end of
+    /// stream.
+    fn write_external_stdout(
+        &self,
+        bytes: Option<Result<Vec<u8>, ShellError>>,
+    ) -> Result<(), ShellError>;
+
+    /// Write some bytes for an `ExternalStream`'s `stderr` stream, or `None` to signal end of
+    /// stream.
+    fn write_external_stderr(
+        &self,
+        bytes: Option<Result<Vec<u8>, ShellError>>,
+    ) -> Result<(), ShellError>;
+
+    /// Write a value for an `ExternalStream`'s `exit_code` stream, or `None` to signal end of
+    /// stream.
+    fn write_external_exit_code(&self, code: Option<Value>) -> Result<(), ShellError>;
+}
+
+/// Implement [StreamDataIo] for the given type. The type is expected to have a shape similar to
+/// `EngineInterfaceImpl` or `PluginInterfaceImpl`. The following struct fields must be defined:
+///
+/// * `read: Mutex<(R, StreamBuffers)>` where `R` implements [`PluginRead`](super::PluginRead)
+/// * `write: Mutex<W>` where `W` implements [`PluginWrite`](super::PluginWrite)
+macro_rules! impl_stream_data_io {
```

Comment:
This macro seems to be used only once. Does it need to be a macro, then? We tend to avoid macros in Nushell since they often make code less readable, but they have their use sometimes.

Reply:
It's used twice, for `EngineInterface` and `PluginInterface`.

I did recently figure out how to do this better with a trait, rather than a macro - like you said, rather avoid huge macros as they're hard to debug.